src/octoprint/util/comm.py

```diff
@@ -3126,134 +3184,213 @@ def _enqueue_for_sending(self, command, linenumber=None, command_type=None, on_s
+						return True
+
+					# now comes the part where we increase line numbers and send stuff - no turning back now
+					self._do_send(command, gcode=gcode)
+
+				# trigger "sent" phase and use up one "ok"
+				if on_sent is not None and callable(on_sent):
+					# we have a sent callback for this specific command, let's execute it now
+					on_sent()
+				self._process_command_phase("sent", command, command_type, gcode=gcode, subcode=subcode, tags=tags)
+
+				# we only need to use up a clear if the command we just sent was either a gcode command or if we also
+				# require ack's for unknown commands
+				use_up_clear = self._unknownCommandsNeedAck
+				if gcode is not None:
+					use_up_clear = True
+
+				if use_up_clear:
+					# if we need to use up a clear, do that now
+					self._clear_to_send.clear()
+
+					if ( self._advanced_ok_detected and self._state == self.STATE_PRINTING):
+						with self._line_mutex:
+							linenumber = self._current_line - 1
+							self._advanced_ok_wait_for_line = linenumber #wait for line by default
+							self._log(">>> LN={} G={} State={}".format(linenumber,gcode,self.getStateString()))
+
+							if( self._advanced_ok_max_line > 0 and not self._send_queue.resend_active):
+								if( gcode in self._advanced_ok_buffered_cmds ):
+									if(linenumber < self._advanced_ok_max_line and self._continue_sending()):
+										self._advanced_ok_wait_for_line = -1 #don't wait, pool instead
+										self._log(">>> continue sending")
+										return True #continue asynchronous sending
+									else:
+										self._advanced_ok_wait_for_line = -1 #don't wait, pool instead
+										return False #stop sending and wait for free buffer
+
+
+				else:
+					# Otherwise we need to tickle the read queue - there might not be a reply
+					# to this command, so our _monitor loop will stay waiting until timeout. We
+					# definitely do not want that, so we tickle the queue manually here
+					self._continue_sending()
+
+			finally:
+				# no matter _how_ we exit this block, we signal that we
+				# are done processing the last fetched queue entry
+				self._send_queue.task_done()
+
+			return False
+		except:
+			self._logger.exception("Caught an exception in the send loop")
+
+
+	def _send_loop(self):
+		"""
+		The send loop is responsible of sending commands in ``self._send_queue`` over the line, if it is cleared for
+		sending (through received ``ok`` responses from the printer's firmware.
+		"""
+
+		self._clear_to_send.wait()
+
+		while self._send_queue_active:
+
+			if self._send_one_from_queue():
+				continue
+
+			if not self._send_queue_active:
+				break
 
-				# now we just wait for the next clear and then start again
+			self._log("####1 wfl={} maxl={} lastl={}".format(self._advanced_ok_wait_for_line, self._advanced_ok_max_line,self._advanced_ok_last_line ))
+			if ( self._advanced_ok_detected
+			and self._advanced_ok_max_line > 0
+			and self._state == self.STATE_PRINTING):
```

Comment:
starting, pausing, cancelling, resuming, finishing? Probably better: `self._state in self.PRINTING_STATES`